charms/knative-serving/src/charm_template.py

```diff
@@ -0,0 +1,248 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+from pathlib import Path
+from typing import List, Union, Optional
+
+import jinja2
+from jinja2 import Environment, FileSystemLoader
+from lightkube import Client, codecs
+from lightkube.core.exceptions import ApiError
+from lightkube.core.resource import NamespacedResource, GlobalResource
+from ops.charm import CharmBase
+from ops.model import ActiveStatus, WaitingStatus, BlockedStatus, MaintenanceStatus
+
+from charm_helpers.exceptions import ErrorWithStatus, LeadershipError
+from lightkube_helpers import check_resources, get_first_worst_error
+
+
+class ExtendedCharmBase(CharmBase):
+    """An extended base class for Charms that includes some conveniences and basic features"""
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.log = logging.getLogger(__name__)
+        self.name = self.model.app.name
+        self.model_name = self.model.name
+
+        self.framework.observe(self.on.config_changed, self.on_config_changed)
+        self.framework.observe(self.on.install, self.on_install)
+        self.framework.observe(self.on.leader_elected, self.on_leader_elected)
+        self.framework.observe(self.on.remove, self.on_remove)
+        self.framework.observe(self.on.upgrade_charm, self.on_upgrade_charm)
+        self.framework.observe(self.on.update_status, self.on_update_status)
+
```

Comment:
Regarding the subscription of handlers to events, do you think there are cases where someone would want to de-subscribe one of the defaults (install, config changed, etc)?  Atm that's awkward, although there's a workaround that achieves almost the same thing (override the default handler with `pass`, eg: `def on_install(...): pass`).  

We could put the actual subscriptions to those default events in a method that could be overridden by the charm author if we think they'd want to really unsubscribe.  What do you think?  Something like:
```
class BaseClass:
  def __init__():
    self.observe_config_changed()
  
  def observe_config_changed():
    self.framework.observe(self.on.config_changed, self.on_config_changed)
```

Or even have a single `subscribe_default_events` that subscribes config, leader, etc all at once.